File: apps/web/src/games/engine/AgeNormalizer.ts
/**
 * AgeNormalizer — z-score normalization of features against age-specific norms
 * Traceability: GAME-FR-011
 *
 * "Each feature is z-score normalized against age-specific norms
 *  (z = (value - age_mean) / age_std) from a normative database tabulated at
 *  0.5-year increments, interpolating between points; features are left
 *  unnormalized with a flag if norms aren't yet available for that age/feature."
 */

/** Single normative entry: mean and std_dev for one feature at one age band */
export interface NormativeEntry {
    age_months_min: number;
    age_months_max: number;
    feature_name: string;
    mean: number;
    std_dev: number;
    is_available: boolean;
}

/** Normative database (loaded from /api/normative or offline cache) */
export type NormativeDatabase = NormativeEntry[];

export type NormalizedFeature = number | null;

export interface NormalizationResult {
    normalized: NormalizedFeature;
    /** true if norms were available; false if left unnormalized per GAME-FR-011 */
    norms_available: boolean;
    mean_used?: number;
    std_dev_used?: number;
}

export class AgeNormalizer {
    private readonly normdb: NormativeDatabase;

    constructor(normdb: NormativeDatabase) {
        this.normdb = normdb;
    }

    /**
     * Normalize a single feature value using z-score with linear interpolation
     * for ages between 0.5-year increments.
     * GAME-FR-011: "interpolating between points"
     */
    normalize(featureName: string, rawValue: number, ageMonths: number): NormalizationResult {
        const params = this.lookupNorms(featureName, ageMonths);

        if (!params || !params.is_available) {
            // GAME-FR-011: "features are left unnormalized with a flag if norms aren't yet available"
            return { normalized: null, norms_available: false };
        }

        const z = (rawValue - params.mean) / params.std_dev;
        return {
            normalized: z,
            norms_available: true,
            mean_used: params.mean,
            std_dev_used: params.std_dev,
        };
    }

    /**
     * Normalize an entire feature record.
     * Returns Record<string, number | null> where null means norms unavailable.
     * GAME-FR-011: null-flagged, not zeroed.
     */
    normalizeAll(
        features: Record<string, number>,
        ageMonths: number,
    ): Record<string, number | null> {
        const result: Record<string, number | null> = {};
        for (const [name, value] of Object.entries(features)) {
            const norm = this.normalize(name, value, ageMonths);
            result[name] = norm.normalized;
        }
        return result;
    }

    /**
     * Look up or interpolate normative parameters for a feature at a given age.
     * GAME-FR-011: "tabulated at 0.5-year increments, interpolating between points"
     */
    private lookupNorms(
        featureName: string,
        ageMonths: number,
    ): { mean: number; std_dev: number; is_available: boolean } | null {
        const relevant = this.normdb.filter((e) => e.feature_name === featureName);
        if (relevant.length === 0) return null;

        // Find the entry whose band contains the age
        const exact = relevant.find(
            (e) => ageMonths >= e.age_months_min && ageMonths < e.age_months_max,
        );

        if (exact) {
            return { mean: exact.mean, std_dev: exact.std_dev, is_available: exact.is_available };
        }

        // Interpolate between adjacent bands
        // Sort by age_months_min
        const sorted = [...relevant].sort((a, b) => a.age_months_min - b.age_months_min);

        // Find surrounding entries for linear interpolation
        let lower: NormativeEntry | null = null;
        let upper: NormativeEntry | null = null;

        for (const entry of sorted) {
            if (entry.age_months_min <= ageMonths) lower = entry;
            if (entry.age_months_min > ageMonths && upper === null) upper = entry;
        }

        if (!lower && !upper) return null;
        if (!lower) return { mean: upper!.mean, std_dev: upper!.std_dev, is_available: upper!.is_available };
        if (!upper) return { mean: lower.mean, std_dev: lower.std_dev, is_available: lower.is_available };

        // Both not available → return unavailable
        if (!lower.is_available || !upper.is_available) {
            return { mean: 0, std_dev: 1, is_available: false };
        }

        // Linear interpolation
        const t =
            (ageMonths - lower.age_months_min) /
            (upper.age_months_min - lower.age_months_min);

        const interpMean = lower.mean + t * (upper.mean - lower.mean);
        const interpStd = lower.std_dev + t * (upper.std_dev - lower.std_dev);

        return { mean: interpMean, std_dev: Math.max(interpStd, 0.001), is_available: true };
    }
}

File: services/ml-service/app/normative/age_normalizer.py
"""
Age normalization for the ML service.
Traceability: ML-FE-002, GAME-FR-011, SRS §8.1

"All features normalized by age against the normative database (GAME-FR-011)"
"Missing-game features are null-flagged rather than zeroed, to avoid bias."

This mirrors the client-side AgeNormalizer.ts but runs server-side for
features that weren't normalized on the client (e.g., computed cross-game features).
"""
import numpy as np
from typing import Dict, Optional, List
import logging

log = logging.getLogger(__name__)

# Normative data structure: {feature_name: {age_band_min: (mean, std)}}
NormativeDB = Dict[str, Dict[float, tuple]]  # age_band_min → (mean, std_dev)


class AgeNormalizer:
    """
    Server-side z-score normalization.
    Traceability: GAME-FR-011, SRS §8.1
    """

    def __init__(self, normative_db: NormativeDB):
        self.db = normative_db

    def normalize(
        self,
        feature_name: str,
        raw_value: float,
        age_months: float,
    ) -> Optional[float]:
        """
        z-score normalize a single feature value.
        Returns None if norms unavailable (GAME-FR-011: null-flag, not zero).
        """
        if np.isnan(raw_value):
            return None

        params = self._lookup(feature_name, age_months)
        if params is None:
            return None  # Null-flagged per GAME-FR-011

        mean, std = params
        if std <= 0:
            return None

        return (raw_value - mean) / std

    def normalize_vector(
        self,
        features: Dict[str, Optional[float]],
        age_months: float,
    ) -> Dict[str, Optional[float]]:
        """
        Normalize all features in a dict.
        Returns None values for features with unavailable norms.
        """
        result = {}
        for name, value in features.items():
            if value is None or (isinstance(value, float) and np.isnan(value)):
                result[name] = None
                continue
            result[name] = self.normalize(name, float(value), age_months)
        return result

    def _lookup(
        self,
        feature_name: str,
        age_months: float,
    ) -> Optional[tuple]:
        """
        Look up normative params with linear interpolation between 0.5-year bands.
        Traceability: GAME-FR-011: "interpolating between points"
        """
        if feature_name not in self.db:
            return None

        bands = self.db[feature_name]
        band_keys = sorted(bands.keys())

        if not band_keys:
            return None

        # Exact match
        for k in band_keys:
            if k <= age_months < k + 6:  # 0.5-year = 6 months
                return bands[k]

        # Below range — use lowest
        if age_months < band_keys[0]:
            return bands[band_keys[0]]

        # Above range — use highest
        if age_months >= band_keys[-1] + 6:
            return bands[band_keys[-1]]

        # Interpolate between adjacent bands
        lower_k = max(k for k in band_keys if k <= age_months)
        upper_k = min(k for k in band_keys if k > age_months)

        lower = bands[lower_k]
        upper = bands[upper_k]

        t = (age_months - lower_k) / (upper_k - lower_k)
        mean = lower[0] + t * (upper[0] - lower[0])
        std = lower[1] + t * (upper[1] - lower[1])

        return (mean, max(std, 1e-6))


def load_normative_db_from_rows(rows: List[dict]) -> NormativeDB:
    """
    Build normative DB from DB query rows.
    Row format: {feature_name, age_months_min, mean, std_dev, is_available}
    """
    db: NormativeDB = {}
    for row in rows:
        if not row.get("is_available", True):
            continue
        fname = row["feature_name"]
        age_min = float(row["age_months_min"])
        if fname not in db:
            db[fname] = {}
        db[fname][age_min] = (float(row["mean"]), float(row["std_dev"]))
    return db
